### PR TITLE
feat(macos): Group B.1 — audio device picker + WAV recording (#237, #239)

### LINKED ISSUE #237: v2: Audio device picker (CoreAudio output routing)
**Parent epic:** #228
**Blocked by:** #234 (M6)

Surface CoreAudio output device selection in the SwiftUI Settings → Audio pane.

The Rust side already supports it: \`AudioSink::set_target(uid)\` and \`list_audio_sinks()\` are implemented in #231 (M3) and consume CoreAudio device UIDs (\`kAudioDevicePropertyDeviceUID\`). The MVP just defaults to \`\"\"\` (system default output) and never calls \`set_target\`.

Work:
- New FFI commands: \`sdr_core_list_audio_devices(out_buf, out_count)\`, \`sdr_core_set_audio_device(uid_utf8)\`. ABI minor bump.
- New event variant: \`SDR_EVT_AUDIO_DEVICE_LIST\` (fires on startup and on hot-plug if we wire the listener).
- SwiftUI Settings → Audio: \`Picker\` with the enumerated devices, \"System default\" as the first option.
- Live re-route on selection — \`AudioSink::set_target\` already handles stop+restart on the sink thread.

Future bonus: \`kAudioHardwarePropertyDefaultOutputDevice\` listener for hot-plug auto-follow. v3.

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Expose two engine features that were fully implemented on the Rust side but had no SwiftUI surface. Follows the project's "no Swift-side reinvention" rule: the engine owns `list_audio_sinks()`, `AudioSink::set_target`, `UiToDsp::Start/StopAudioRecording`, and `WavWriter`; this PR is FFI plumbing + SwiftUI wiring.

Closes #237 and #239.

## Engine audit (why no Swift-side logic)

| Feature | Engine source of truth | Swift work |
|---|---|---|
| Device enumeration | `sdr-sink-audio::list_audio_sinks()` | FFI wrap, render in Picker |
| Device routing | `AudioSink::set_target` (transactional swap + rollback) | FFI wrap, persist UID in UserDefaults |
| WAV recording | `UiToDsp::Start/StopAudioRecording` + `WavWriter` | FFI wrap, generate timestamped path |
| Recording confirm | `DspToUi::AudioRecording{Started,Stopped}` | New event variants, flip engine-authoritative model field |
| Failure surfaces | `DspToUi::Error(...)` (existing) | No-op — existing error event arm handles it |

## What changed

**sdr-ffi (ABI 0.3 → 0.4, additive):**
- `sdr_core_audio_device_count` / `_name` / `_uid` — handle-free, caller-allocated buffers matching the existing `sdr_core_device_name` contract.
- `sdr_core_set_audio_device(uid)` — empty string = system default.
- `sdr_core_start_audio_recording(path)` / `_stop_audio_recording`.
- `SDR_EVT_AUDIO_RECORDING_STARTED` (path payload) / `_STOPPED`. Removed from the "silently dropped" fallback; new payload struct added to the union.
- Header updated; `make ffi-header-check` clean.

**SdrCoreKit:**
- `SdrCore.AudioDevice` + static `audioDevices` / `audioDeviceName(at:)` / `audioDeviceUid(at:)`.
- `setAudioDevice`, `startAudioRecording(path:)`, `stopAudioRecording` command wrappers.
- `SdrCoreEvent.audioRecordingStarted(path:)` / `audioRecordingStopped` variants.

**SDRMac:**
- `CoreModel`: `audioDevices`, `selectedAudioDeviceUid` (persisted via `UserDefaults`, restored on bootstrap, re-applied through `syncToEngine`), `audioRecordingPath` (engine-authoritative — flipped only by events).
- **Settings → Audio pane**: real device `Picker` replacing the placeholder. Refreshes on appear so CoreAudio hot-plugs (Bluetooth / USB) show up without app restart.
- **New sidebar Recording section**: toggle + filename + reveal-in-Finder button. Default path `~/Documents/SDRMac/Audio/sdr-audio-YYYYMMDD-HHMMSS.wav`; directory created lazily.

## Tests

- 9 new `sdr-ffi` unit tests: null-handle rejection, null/empty-path rejection, UTF-8 round-trips, out-of-range indices, start/stop through a live engine handle.
- `cargo test --workspace` green.
- `cargo clippy --workspace --all-targets -- -D warnings` clean.
- `make ffi-header-check` clean.
- Release Mac app (`make mac-app`) built and smoke-tested on live hardware: Picker enumerates + switches devices with audio routing to the new device; Recording toggle writes a playable WAV to `~/Documents/SDRMac/Audio/`.

## Test plan

- [x] Unit tests pass (`cargo test --workspace`)
- [x] Clippy clean (`cargo clippy --workspace --all-targets -- -D warnings`)
- [x] Header drift lint clean (`make ffi-header-check`)
- [x] Release build of Mac app succeeds (`make mac-app`)
- [x] Settings → Audio shows enumerated devices, switching routes audio
- [x] Sidebar Recording toggle starts + stops; path shown + reveal-in-Finder works
- [x] Persisted device UID survives app restart (UserDefaults)
- [ ] CodeRabbit review

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Pick and persist an audio output device in Settings (includes “System default”); device list refreshes on appear and selection restored on launch.
  * Start/stop WAV audio recording from the sidebar; files saved to Documents/SDRMac/Audio with unique timestamped names and a Finder “reveal” button.
  * Sidebar shows live, engine-confirmed recording state.

* **Bug Fixes**
  * Prevents creating a started-recording state with an empty path; recording toggle guarded against rapid re-entrant changes.

### CODE REVIEW

File: apps/macos/Packages/SdrCoreKit/Sources/SdrCoreKit/SdrCoreEvent.swift [hunk 1]
```diff
@@ -116,6 +127,14 @@ public enum SdrCoreEvent: Sendable, Equatable {
             guard let cstr else { return .error("") }
             return .error(String(cString: cstr))
 
+        case kAudioRecordingStarted:
+            let cstr = payload.audio_recording.path_utf8
+            guard let cstr else { return .audioRecordingStarted(path: "") }
+            return .audioRecordingStarted(path: String(cString: cstr))
```
Comment (jasonherald) [inline apps/macos/Packages/SdrCoreKit/Sources/SdrCoreKit/SdrCoreEvent.swift]: Fixed in 0b2a3ff — null path pointer now returns `nil` from `fromC` so the event is silently dropped, with a comment explaining why empty-path-started would be worse than skipping the event entirely. The eventual `.audioRecordingStopped` still clears state correctly.

File: apps/macos/SDRMac/Views/RecordingSection.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    /// sortable, human-readable, and unique enough that back-to-
+    /// back recordings don't collide.
+    static func generateRecordingPath() -> String {
+        let dir = SDRMacApp.audioRecordingsDirectory()
+        let formatter = DateFormatter()
+        formatter.locale = Locale(identifier: "en_US_POSIX")
+        formatter.dateFormat = "yyyyMMdd-HHmmss"
+        let stamp = formatter.string(from: Date())
+        return dir.appendingPathComponent("sdr-audio-\(stamp).wav").path
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply apps/macos/SDRMac/Views/RecordingSection.swift:126]: Fixed in 0b2a3ff — applied your suggested patch verbatim: added `-SSS` milliseconds to the timestamp and a `while FileManager.fileExists` loop with an integer suffix. Millisecond-precise timestamps make organic collisions vanishingly unlikely; the suffix handles the edge case where a user replays an old recording or a previous-session file is still around.

File: include/sdr_core.h [hunk 3]
```diff
@@ -148,6 +148,49 @@ int32_t sdr_core_device_name(
...
+    uint32_t index,
+    char*    out_buf,
+    size_t   buf_len
+);
+
+int32_t sdr_core_audio_device_uid(
+    uint32_t index,
+    char*    out_buf,
+    size_t   buf_len
+);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply include/sdr_core.h:206]: Agreed on the underlying race — fixed in 0b2a3ff.

Went with a **thread-local snapshot** instead of a new snapshot-handle ABI:

- `sdr_core_audio_device_count` refreshes a per-thread `Vec<AudioDevice>` and returns its length.
- `sdr_core_audio_device_name(i)` / `_uid(i)` read from that snapshot, so for a given thread every index pairs the right name with the right UID even if a device hot-plugs mid-enumeration.
- Callers that skip `_count` (one-shot lookups) lazy-refresh on first access; they don't get multi-index consistency but single-index calls still work.
- No new ABI surface — the existing Swift `SdrCore.audioDevices` computed property on this PR gets atomic enumeration for free without any caller changes.

Rationale for TLS over a snapshot-handle API: hosts already use this pattern for `sdr_core_device_count` / `_name` without a handle, the snapshot scope matches how SwiftUI pickers actually enumerate (synchronous loop on the main actor), and it keeps the ABI 0.4 surface small. The updated header doc now spells out the atomicity contract, and two new regression tests (`audio_device_name_and_uid_share_snapshot_after_count`, `audio_device_lazy_refresh_when_count_not_called`) pin it down.

File: apps/macos/Packages/SdrCoreKit/Sources/SdrCoreKit/SdrCoreEvent.swift [hunk 4]
```diff
@@ -116,6 +127,21 @@ public enum SdrCoreEvent: Sendable, Equatable {
...
+            // Drop the event on a null path pointer rather than
+            // emit `.audioRecordingStarted(path: "")` — an empty
+            // path would flip CoreModel's `audioRecordingPath`
+            // into a bogus non-nil "recording" state with no file
+            // the UI can meaningfully display or reveal. The
+            // subsequent `.audioRecordingStopped` will still
+            // clear state correctly. Per CodeRabbit round 1 on
+            // PR #344.
+            guard let cstr = payload.audio_recording.path_utf8 else { return nil }
+            return .audioRecordingStarted(path: String(cString: cstr))
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply apps/macos/Packages/SdrCoreKit/Sources/SdrCoreKit/SdrCoreEvent.swift]: Fixed in 39c40fd — applied your suggested patch. The guard now rejects both null pointers AND zero-length strings, so the only events that fire are ones with a real filesystem path the UI can display + reveal.

File: apps/macos/SDRMac/Views/RecordingSection.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                get: { recording },
+                set: { on in
+                    if on {
+                        let path = Self.generateRecordingPath()
+                        model.startAudioRecording(to: path)
+                    } else {
+                        model.stopAudioRecording()
+                    }
+                }
+            ))
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply apps/macos/SDRMac/Views/RecordingSection.swift:50]: Fixed in 39c40fd — added a `pendingTransition` `@State` flag + `.disabled`. Behavior:

- On toggle, if already pending → swallow the event. Otherwise flip pending=true and fire the command.
- `.onChange(of: model.audioRecordingPath)` clears pending once the engine confirms the transition.
- Fallback: `.onChange(of: model.lastError)` also clears pending if a non-nil error lands while we're pending. This handles the `start` failure path (engine emits `.error` without ever flipping `audioRecordingPath`) — without it the toggle would stay locked forever after a failed start.

The `start`-then-immediate-`stop` case still works correctly since both transitions end up flipping the path (nil → some → nil).

File: crates/sdr-ffi/src/enumerate.rs [hunk 6]
```diff
@@ -117,6 +118,249 @@ pub unsafe extern "C" fn sdr_core_device_name(
...
+    std::panic::catch_unwind(|| {
+        let len = refresh_audio_device_snapshot();
+        // Cap at u32::MAX defensively. In practice there are <100
+        // devices on any real system.
+        u32::try_from(len).unwrap_or(u32::MAX)
+    })
+    .unwrap_or_else(|_| {
+        set_last_error("sdr_core_audio_device_count: panic during enumeration");
+        0
+    })
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ffi/src/enumerate.rs:248]: Fixed in 39c40fd — applied your suggested patch. Also added a regression test (`audio_device_count_clears_stale_last_error`) that poisons the thread-local via a null-buffer `_name` call and then verifies the next successful `_count` clears it. All 54 sdr-ffi tests still pass.

File: crates/sdr-ffi/src/enumerate.rs [hunk 7]
```diff
@@ -169,4 +419,169 @@ mod tests {
...
+            .expect("FFI writes a NUL at `written` per contract");
+
+        // UID for index 0 is typically the empty string ("system default"),
+        // but that's backend policy — we just check the call works.
+        let mut uid_buf = [0_u8; 256];
+        let rc = unsafe {
+            sdr_core_audio_device_uid(0, uid_buf.as_mut_ptr().cast::<c_char>(), uid_buf.len())
+        };
+        assert!(rc >= 0, "expected success, got {rc}");
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ffi/src/enumerate.rs:654]: Fixed in c7927a6 — named constants at the top of the test module:

- `REJECTION_BUF_LEN = 64` (null / zero-length / out-of-range probe buffers)
- `REJECTION_INNER_BUF_LEN = 32` (the nonzero `buf_len` paired with a null `out_buf`, documented inline so future readers know it exists specifically to hit the null branch and not the `buf_len == 0` branch)
- `AUDIO_BUF_LEN = 256` (round-trip + snapshot buffers — covers CoreAudio display names and decimal-`AudioDeviceID` UIDs)
- `MAX_REASONABLE_AUDIO_DEVICE_COUNT = 1024` (sanity upper bound on `_count`)

Scoped to the new audio_device_* tests I added in this PR — didn't touch the pre-existing rtl-sdr tests to keep the diff minimal. All 54 sdr-ffi tests still green.

File: crates/sdr-ffi/src/enumerate.rs [hunk 8]
```diff
@@ -117,6 +118,255 @@ pub unsafe extern "C" fn sdr_core_device_name(
...
+        // a successful refresh. Per CodeRabbit round 2 on PR #344.
+        clear_last_error();
+        // Cap at u32::MAX defensively. In practice there are <100
+        // devices on any real system.
+        u32::try_from(len).unwrap_or(u32::MAX)
+    })
+    .unwrap_or_else(|_| {
+        set_last_error("sdr_core_audio_device_count: panic during enumeration");
+        0
+    })
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ffi/src/enumerate.rs:248]: Fixed in 88f9ac5 — applied your suggested patch. The `unwrap_or_else` branch now clears `AUDIO_DEVICE_SNAPSHOT` before setting the last-error and returning 0, so a stale pre-panic snapshot can't silently disagree with the zero count.

No regression test: reliably forcing `list_audio_sinks()` to panic would need mocking the backend, and the fix is 4 lines of defensive clearing inside the existing unwrap_or_else — inline comment captures the invariant. 54 sdr-ffi tests still green.

File: crates/sdr-ffi/src/enumerate.rs [hunk 9]
```diff
@@ -169,4 +430,202 @@ mod tests {
...
+            let rc_uid = unsafe {
+                sdr_core_audio_device_uid(i, uid_buf.as_mut_ptr().cast::<c_char>(), uid_buf.len())
+            };
+            assert!(rc_name >= 0);
+            assert!(rc_uid >= 0);
+            // Both calls must succeed for every index the count
+            // call reported — if `_uid` returned Device error while
+            // `_name` succeeded, that's the inconsistency the
+            // thread-local is meant to prevent.
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ffi/src/enumerate.rs:607]: Fixed in 058d471 — applied your suggested patch. The test now clones `AUDIO_DEVICE_SNAPSHOT` into an `expected` vec right after `_count`, asserts `expected.len()` matches the u32 count, then for each index compares the FFI-returned display name + UID strings byte-for-byte against `dev.display_name` / `dev.node_name`. Any future refactor that re-introduced a per-call `list_audio_sinks()` round-trip in `_name` / `_uid` would fail this test loudly.

54 sdr-ffi tests still green.

File: crates/sdr-ffi/src/enumerate.rs [hunk 10]
```diff
@@ -117,6 +118,266 @@ pub unsafe extern "C" fn sdr_core_device_name(
...
+        // straight to `_name(0)` without calling `_count` still
+        // gets the right answer. This does a backend query per
+        // call in that path; the atomic multi-index case still
+        // uses the cached snapshot.
+        let idx = usize::try_from(index).unwrap_or(usize::MAX);
+        AUDIO_DEVICE_SNAPSHOT.with(|cell| {
+            if cell.borrow().is_empty() {
+                let fresh = sdr_sink_audio::list_audio_sinks();
+                *cell.borrow_mut() = fresh;
+            }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ffi/src/enumerate.rs]: Fixed in 954f251 — applied your suggested patch. The lazy path now calls `refresh_audio_device_snapshot()` instead of inlining list + assign, so the `_count` path and the lazy path share one source of truth for how the thread-local gets populated. Two nested `.with` calls instead of one, but the drift risk is gone.

54 sdr-ffi tests still green.
